### ISSUE TITLE
ComputationGraph persisting problems

### ISSUE THREAD

**Issue by AlexDmitriychuk** [deeplearning4j/deeplearning4j#4826]
#### Issue Description
encountered:
Persisted computation graph with updater restart fitting after load from file.
expected:
continue fitting

#### Version Information

dl4j version: 0.9.1
platform: linux 

ANN based on dl4j example CnnSentenceClassificationExample

logs:
> Resuming last session
> Operation: train
> Suboperation: conv-classifier
> alpha_train.csv -> conv-model.zip.145
> Train started.
> 2018-03-22 09:00:00 INFO  org.nd4j.linalg.factory.Nd4jBackend - Loaded [CpuBackend] backend
> 2018-03-22 09:00:00 INFO  org.nd4j.nativeblas.NativeOpsHolder - Number of threads used for NativeOps: 4
> 2018-03-22 09:00:01 INFO  org.nd4j.nativeblas.Nd4jBlas - Number of threads used for BLAS: 4
> 2018-03-22 09:00:01 INFO  o.n.l.a.o.e.DefaultOpExecutioner - Backend used: [CPU]; OS: [Linux]
> 2018-03-22 09:00:01 INFO  o.n.l.a.o.e.DefaultOpExecutioner - Cores: [8]; Memory: [14.0GB];
> 2018-03-22 09:00:01 INFO  o.n.l.a.o.e.DefaultOpExecutioner - Blas vendor: [OPENBLAS]
> Resuming model: conv-model.zip.145
> Resuming iteration: 146
> 2018-03-22 09:00:04 INFO  o.d.nn.graph.ComputationGraph - Starting ComputationGraph with WorkspaceModes set to [training: SINGLE; inference: SINGLE]
> 2018-03-22 09:00:05 INFO  o.d.ui.play.PlayUIServer - DL4J UI Server started at http://localhost:9000
> 2018-03-22 09:00:05 INFO  o.d.ui.play.PlayUIServer - StatsStorage instance attached to UI: InMemoryStatsStorage(uid=223dc713)
> Started at Thu Mar 22 09:00:05 EET 2018, look at http://localhost:9000/train/system
> 
> ============================================================================================================================================
> VertexName (VertexType)                 nIn,nOut       TotalParams    ParamsShape                    Vertex Inputs
> ============================================================================================================================================
> input (InputVertex)                     -,-            -              -                              -
> cnn1 (ConvolutionLayer)                 1,18           630            b:{1,18}, W:{18,1,1,34}        [input]
> cnn2 (ConvolutionLayer)                 1,17           1173           b:{1,17}, W:{17,1,1,68}        [input]
> cnn3 (ConvolutionLayer)                 1,16           1648           b:{1,16}, W:{16,1,1,102}       [input]
> cnn4 (ConvolutionLayer)                 1,15           2055           b:{1,15}, W:{15,1,1,136}       [input]
> cnn5 (ConvolutionLayer)                 1,14           2394           b:{1,14}, W:{14,1,1,170}       [input]
> cnn6 (ConvolutionLayer)                 1,13           2665           b:{1,13}, W:{13,1,1,204}       [input]
> merge (MergeVertex)                     -,-            -              -                              [cnn1, cnn2, cnn3, cnn4, cnn5, cnn6]
> pool (GlobalPoolingLayer)               -,-            0              -                              [merge]
> dense1 (DenseLayer)                     93,373         35062          b:{1,373}, W:{93,373}          [pool]
> dense2 (DenseLayer)                     373,186        69564          b:{1,186}, W:{373,186}         [dense1]
> out (OutputLayer)                       186,3          561            b:{1,3}, W:{186,3}             [dense2]
> --------------------------------------------------------------------------------------------------------------------------------------------
>             Total Parameters:  115752
>         Trainable Parameters:  115752
>            Frozen Parameters:  0
> ============================================================================================================================================
> 
> Evaluate model....
> 
> Examples labeled as 0 classified by model as 0: 873 times
> Examples labeled as 0 classified by model as 1: 292 times
> Examples labeled as 0 classified by model as 2: 5 times
> Examples labeled as 1 classified by model as 0: 25 times
> Examples labeled as 1 classified by model as 1: 83948 times
> Examples labeled as 1 classified by model as 2: 23 times
> Examples labeled as 2 classified by model as 0: 2 times
> Examples labeled as 2 classified by model as 1: 23 times
> Examples labeled as 2 classified by model as 2: 1647 times
> 
> 
> ==========================Scores========================================
>  # of classes:    3
>  Accuracy:        0.9957
>  Precision:       0.9832
>  Recall:          0.9102
>  F1 Score:        0.9418
> Precision, recall & F1: macro-averaged (equally weighted avg. of 3 classes)
> ========================================================================
> Epoch: 146
> Iteration time: 45sec, Average: 45sec
> Total time spent: 109min
> Estimate time: 640min
> Epoch: 147
> Iteration time: 117sec, Average: 117sec
> Total time spent: 286min
> Estimate time: 1663min
> Epoch: 148
> Iteration time: 62sec, Average: 62sec
> Total time spent: 153min
> Estimate time: 880min
> Epoch: 149
> Iteration time: 60sec, Average: 60sec
> Total time spent: 149min
> Estimate time: 851min
> Epoch: 150
> Iteration time: 68sec, Average: 68sec
> Total time spent: 170min
> Estimate time: 963min
> Model saved to conv-model.zip.150
> Evaluate model....
> 
> Examples labeled as 0 classified by model as 0: 1170 times
> Examples labeled as 1 classified by model as 0: 83996 times
> Examples labeled as 2 classified by model as 0: 1672 times
> 
> Warning: 2 classes were never predicted by the model and were excluded from average precision
> Classes excluded from average precision: [1, 2]
> 
> ==========================Scores========================================
>  # of classes:    3
>  Accuracy:        0.0135
>  Precision:       0.0135	(2 classes excluded from average)
>  Recall:          0.3333
>  F1 Score:        0.0266	(2 classes excluded from average)
> Precision, recall & F1: macro-averaged (equally weighted avg. of 3 classes)
> ========================================================================
> Epoch: 151

**Comment by AlexDBlack** [comment]
Can you please provide some code to reproduce this? Something isolated that I can easily run would be great. 
Unfortunately there isn't enough enough information here to identify the cause of the problem and reproduce it.
Thanks!

**Comment by AlexDmitriychuk** [comment]
ok, I've created the patch for dl4j example https://github.com/deeplearning4j/dl4j-examples/blob/master/dl4j-examples/src/main/java/org/deeplearning4j/examples/convolution/sentenceclassification/CnnSentenceClassificationExample.java

[compGraphTestPatch.patch.txt](https://github.com/deeplearning4j/deeplearning4j/files/1837468/compGraphTestPatch.patch.txt)

**Comment by AlexDBlack** [comment]
Thus far I'm unable to reproduce any sort of a problem here - though I haven't exactly run your unmodified code (I don't exactly have 10 hours to debug this here...) - I had to cut down the number of epochs.

Now, it *looks* like you are training for some number of epochs (your code has N=20 epochs, your comments have 151 epochs total?) saving + restoring the model, then training for N more epochs.
Then you are comparing the model at N epochs vs. 2N epochs, and wondering why they are different?

If that's the case - then there's a million reasons why a model could perform poorly with more training. Most notably, training can become unsable after some time (perhaps due to insufficient regularization), or you can have severe overfitting.
That seems like the most likely cause here. Model persistance is pretty well tested at this point.

However, if I've misinterpreted your code - and you are training the original and restored models *for the exact same total number of epochs for both models* - then I'll re-open this and take another look.

**Comment by AlexDmitriychuk** [comment]
Actually model I use learn further if I don't restart it, so learning process reached epoch 215.
So now I can't persist model and continue training after restoring from file. Nice feature but it is not working. 

> Epoch: 215
> Iteration time: 220sec, Average: 220sec
> Total time spent: 791min
> Estimate time: 2878min
> Model saved to conv-model.zip.215
> Evaluate model....

> Examples labeled as 0 classified by model as 0: 1162 times
> Examples labeled as 0 classified by model as 1: 5 times
> Examples labeled as 0 classified by model as 2: 3 times
> Examples labeled as 1 classified by model as 0: 10 times
> Examples labeled as 1 classified by model as 1: 83985 times
> Examples labeled as 1 classified by model as 2: 1 times
> Examples labeled as 2 classified by model as 0: 7 times
> Examples labeled as 2 classified by model as 2: 1665 times
> 
> 
> Scores
>   of classes:    3
>  Accuracy:        0.9997
>  Precision:       0.9944
>  Recall:          0.9963
>  F1 Score:        0.9953
> Precision, recall & F1: macro-averaged (equally weighted avg. of 3 classes)
>
>

**Comment by AlexDBlack** [comment]
> Actually model I use learn further if I don't restart it, so learning process reached epoch 215.
So now I can't persist model and continue training after restoring from file. Nice feature but it is not working.

To be blunt: I'm not seeing an issue here that would indicate a bug. Though to be honest it's hard to decypher exactly what the expected behaviour is from your code.

If you suspect there's a legitimate issue here, here's what I'd like to I can reproduce it:
Give me some code that does something along the lines of the following:
(a) Trains the net for X epochs/iterations
(b) Saves the net
(c) loads the net
(d) trains both the original, and the serialized/loaded nets in exactly the same way

Randomization in the data pipeline aside, the two networks should have identical state at the end of training.
